kiva/quartz/ABCGI.pyx

```diff
@@ -1523,30 +1523,29 @@ cdef class CGBitmapContext(CGContext):
     def width(self):
         return CGBitmapContextGetWidth(self.context)
 
-    def __getsegcount__(self, void* tmp):
-        cdef int *lenp
-        lenp = <int *>tmp
-        if lenp != NULL:
-            lenp[0] = self.height()*self.bytes_per_row
-        return 1
-
-    def __getreadbuffer__(self, int segment, void** ptr):
-        # ignore invalid segment; the caller can't mean anything but the only
-        # segment available; we're all adults
-        ptr[0] = self.data
-        return self.height()*self.bytes_per_row
-
-    def __getwritebuffer__(self, int segment, void** ptr):
-        # ignore invalid segment; the caller can't mean anything but the only
-        # segment available; we're all adults
-        ptr[0] = self.data
-        return self.height()*self.bytes_per_row
-
-    def __getcharbuffer__(self, int segment, char** ptr):
-        # ignore invalid segment; the caller can't mean anything but the only
-        # segment available; we're all adults
-        ptr[0] = <char*>(self.data)
-        return self.height()*self.bytes_per_row
+    def __getbuffer__(self, Py_buffer *buffer, int flags):
+        cdef Py_ssize_t shape[2]
+        cdef Py_ssize_t strides[2]
+
+        shape[0] = self.bytes_per_row
+        shape[1] = self.height()
+        strides[0] = self.bytes_per_row
+        strides[1] = 1
+
+        buffer.buf = <char *>(self.data)
+        buffer.format = 'b'
+        buffer.internal = NULL
+        buffer.itemsize = 1
+        buffer.len = self.height() * self.bytes_per_row
+        buffer.ndim = 2
+        buffer.obj = self
+        buffer.readonly = 1
+        buffer.shape = shape
+        buffer.strides = strides
+        buffer.suboffsets = NULL
+
+    def __releasebuffer__(self, Py_buffer *buffer):
+        pass
```

Comment:
I lacked any context on these changes so I just tried to follow along with this https://cython.readthedocs.io/en/latest/src/userguide/buffer.html 

From what I can see changes LGTM.  I'm guessing we don't have a need to do the reference count stuff in `__releasebuffer__` here?

Reply:
Thanks for pointing this out. I re-read the Cython docs again and you're right that we don't need to worry about reference counting. We don't mutate the data in a way that is dangerous to buffers created here, so we don't need any special considerations in `__releasebuffer__`.

That said, I've made a small change to use dynamically allocated memory for `shape` and `strides` since the first version (with stack-allocated arrays) was likely quite fragile.